src/graphics/aurora/flashfont.cpp

```diff
@@ -0,0 +1,208 @@
+/* xoreos - A reimplementation of BioWare's Aurora engine
+ *
+ * xoreos is the legal property of its developers, whose names
+ * can be found in the AUTHORS file distributed with this source
+ * distribution.
+ *
+ * xoreos is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * xoreos is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with xoreos. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file
+ *  Class for rendering swf fonts.
+ */
+
+#include "src/common/bitstream.h"
+#include "src/common/encoding.h"
+#include "src/common/rect.h"
+
+#include "src/aurora/gfxfile.h"
+
+#include "src/graphics/aurora/flashfont.h"
+
+#ifdef ENABLE_CAIRO
+
+static cairo_user_data_key_t kUserDataKey;
```

Comment:
Why does that need to be a global?

Reply:
Because the methods renderGlyph and textToGlyphs are static and need a key to get the proper FlashFont class which is set as user data in a cairo_font_face_t. But it can be made const.

Reply:
But this basically equivalent to having a global `FlashFont *`.

I see why you're doing this, because the glyph callback needs to know about the font, and the usual Cairo method is by going through the user data (instead of, for example, a `void *` parameter).

However, the way you're doing this, this means you can ever only have one Font at the same time.

If it were an integer, I would have suggested `Common::generateIDNumber()`, but Cairo does it in a weird way, by using the memory address.

So you should probably rather have (non-static) `cairo_user_data_key_t` member variable in FlashFont, and use that.

(This might still clash if we're using Cairo from different threads for certain thread implementations, but the probability there is low)

Reply:
I think you misunderstand something here. The problem with your suggested approach is, that renderGlyph and textToGlyph are static and don't have a direct reference or pointer to the specific FlashFont object. The FlashFont class is for example extracted in line 80 to 85 in flashfont.cpp using the user data key, meaning i cannot get access to the user data without the key. The key associated user data is unique in every font as written [https://www.cairographics.org/manual/cairo-cairo-font-face-t.html#cairo-font-face-set-user-data](here), meaning you could use for every FlashFont object set as user data of a cairo_font_face_t the same key since every cairo_font_face_t contains exactly one FlashFont. I know that this solution is not ideal, but i cannot think of a better approach for this c-style callbacks.

Reply:
Aaah, okay, if the key is unique to every font, yeah, that would work.

That's just...very weird of Cairo.